File: apps/mobile/src/services/api/contracts.ts
import type { Item, PlaceDetails, PlacePath, PlaceSummary } from "./types"

export function isPlaceSummary(value: unknown): value is PlaceSummary {
  if (!isRecord(value)) return false

  return (
    isString(value.id) &&
    isString(value.name) &&
    isNullableString(value.photoUrl) &&
    isNullableString(value.description) &&
    isNonNegativeNumber(value.childPlaceCount) &&
    isNonNegativeNumber(value.itemCount) &&
    isString(value.createdAt) &&
    isString(value.updatedAt)
  )
}

export function isPlaceDetails(value: unknown): value is PlaceDetails {
  if (!isRecord(value)) return false

  return (
    isString(value.id) &&
    isString(value.name) &&
    isNullableString(value.parentPlaceId) &&
    isNullableString(value.photoUrl) &&
    isNullableString(value.description) &&
    isString(value.createdAt) &&
    isString(value.updatedAt) &&
    Array.isArray(value.ancestry) &&
    value.ancestry.every(isPlacePath) &&
    Array.isArray(value.children) &&
    value.children.every(isPlaceSummary) &&
    Array.isArray(value.items) &&
    value.items.every(isItem)
  )
}

export function isItem(value: unknown): value is Item {
  if (!isRecord(value)) return false

  return (
    isString(value.id) &&
    isString(value.placeId) &&
    isString(value.name) &&
    isString(value.iconKey) &&
    isNullableString(value.photoUrl) &&
    isNullableString(value.brand) &&
    isNullableString(value.model) &&
    isNullableString(value.serialNumber) &&
    isNullableString(value.category) &&
    isNullableString(value.productionDate) &&
    isNullableString(value.purchaseDate) &&
    (value.purchasePrice === null || typeof value.purchasePrice === "number") &&
    isNullableString(value.purchaseCurrency) &&
    isNullableString(value.warrantyUntil) &&
    isNullableString(value.productUrl) &&
    typeof value.quantity === "number" &&
    value.quantity > 0 &&
    Array.isArray(value.tags) &&
    value.tags.every(isString) &&
    isNullableString(value.notes) &&
    isString(value.createdAt) &&
    isString(value.updatedAt)
  )
}

export function isArrayOf<T>(guard: (value: unknown) => value is T) {
  return (value: unknown): value is T[] => Array.isArray(value) && value.every(guard)
}

function isPlacePath(value: unknown): value is PlacePath {
  return isRecord(value) && isString(value.id) && isString(value.name)
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value)
}

function isString(value: unknown): value is string {
  return typeof value === "string"
}

function isNullableString(value: unknown): value is string | null {
  return value === null || isString(value)
}

function isNonNegativeNumber(value: unknown): value is number {
  return typeof value === "number" && value >= 0
}

File: apps/backend/Yze.Api/Features/Contracts.cs
namespace Yze.Api.Features;

public sealed record CreatePlaceRequest(
    string? Name,
    Guid? ParentPlaceId,
    string? PhotoUrl,
    string? Description);

public sealed record UpdatePlaceRequest(string? Name, string? PhotoUrl, string? Description);

public sealed record MovePlaceRequest(Guid? ParentPlaceId);

public sealed record PlacePathResponse(Guid Id, string Name);

public sealed record PlaceSummaryResponse(
    Guid Id,
    string Name,
    string? PhotoUrl,
    string? Description,
    int ChildPlaceCount,
    int ItemCount,
    DateTimeOffset CreatedAt,
    DateTimeOffset UpdatedAt);

public sealed record PlaceDetailsResponse(
    Guid Id,
    bool IsRoot,
    string Name,
    Guid? ParentPlaceId,
    string? PhotoUrl,
    string? Description,
    DateTimeOffset CreatedAt,
    DateTimeOffset UpdatedAt,
    IReadOnlyList<PlacePathResponse> Ancestry,
    IReadOnlyList<PlaceSummaryResponse> Children,
    IReadOnlyList<ItemResponse> Items);

public sealed record OrganizerRootResponse(
    Guid Id,
    string Name,
    int ChildPlaceCount,
    int ItemCount);

public sealed record OrganizerPlaceResponse(
    Guid Id,
    Guid ParentPlaceId,
    string Name,
    string? PhotoUrl,
    string? Description,
    DateTimeOffset CreatedAt,
    DateTimeOffset UpdatedAt);

public sealed record OrganizerItemResponse(
    Guid Id,
    Guid PlaceId,
    string Name,
    string IconKey,
    int Quantity);

public sealed record OrganizerTreeResponse(
    OrganizerRootResponse Root,
    IReadOnlyList<OrganizerPlaceResponse> Places,
    IReadOnlyList<OrganizerItemResponse> Items);

public sealed record ItemWriteRequest(
    string? Name,
    string? IconKey,
    string? PhotoUrl,
    string? Brand,
    string? Model,
    string? SerialNumber,
    string? Category,
    DateOnly? ProductionDate,
    DateOnly? PurchaseDate,
    decimal? PurchasePrice,
    string? PurchaseCurrency,
    DateOnly? WarrantyUntil,
    string? ProductUrl,
    int? Quantity,
    IReadOnlyList<string>? Tags,
    string? Notes);

public sealed record MoveItemRequest(Guid? PlaceId);

public sealed record ItemResponse(
    Guid Id,
    Guid PlaceId,
    string Name,
    string IconKey,
    string? PhotoUrl,
    string? Brand,
    string? Model,
    string? SerialNumber,
    string? Category,
    DateOnly? ProductionDate,
    DateOnly? PurchaseDate,
    decimal? PurchasePrice,
    string? PurchaseCurrency,
    DateOnly? WarrantyUntil,
    string? ProductUrl,
    int Quantity,
    IReadOnlyList<string> Tags,
    string? Notes,
    DateTimeOffset CreatedAt,
    DateTimeOffset UpdatedAt);
